File: packages/checks/src/checks/noMonomorphicStructGet.ts
import { Array, Option, pipe, Struct, flow } from "effect"
import * as ts from "typescript"
import { hasExportModifier, unwrapTransparentExpression } from "./support/tsNode.js"
import type { CheckContext } from "@better-typescript/core/engine/check/data"
import type { Detection } from "@better-typescript/core/engine/location/data"
import { makeCheck } from "../defineCheck.js"
import { makeDetection } from "@better-typescript/core/engine/check"
import { strictEqual } from "@better-typescript/core/engine/equivalence"

const message = "Avoid monomorphizing Struct.get at its declaration."

const hint =
  "Keep Struct.get polymorphic. Inline it at a typed consumer, or put the " +
  "domain type on the consuming value or result rather than on the getter."

const effectPackagePathSegments: ReadonlyArray<string> = Array.make(
  "/node_modules/effect/",
  "/node_modules/@effect/"
)

const structModuleSuffixes: ReadonlyArray<string> = Array.make("/Struct.d.ts", "/Struct.ts")

const declarationIsEffectStructModule = (declaration: ts.Declaration) => {
  const sourceFile = declaration.getSourceFile()
  const fileName = sourceFile.fileName.replaceAll("\\", "/")
  const pathIncludesSegment = (segment: string) => fileName.includes(segment)
  const pathEndsWithSuffix = (suffix: string) => fileName.endsWith(suffix)
  const inEffectPackage = Array.some(effectPackagePathSegments, pathIncludesSegment)
  const isStructModule = Array.some(structModuleSuffixes, pathEndsWithSuffix)
  const effectStructModuleConditions = Array.make(inEffectPackage, isStructModule)

  return Array.every(effectStructModuleConditions, Boolean)
}

const symbolDeclaredInEffectStructModule = (symbol: ts.Symbol) => {
  const declarations = symbol.getDeclarations() ?? Array.empty()

  return Array.some(declarations, declarationIsEffectStructModule)
}

const monomorphicStructGetMatches = (context: CheckContext) => {
  const checker = context.checker
  const match = makeDetection(context)

  const declarationIsExported = (declaration: ts.VariableDeclaration) =>
    pipe(
      Option.some(declaration.parent.parent),
      Option.filter(ts.isVariableStatement),
      Option.exists(hasExportModifier)
    )

  const typeNodeIsNonGenericCallable = (typeNode: ts.TypeNode) => {
    const declaredType = checker.getTypeFromTypeNode(typeNode)
    const signatures = declaredType.getCallSignatures()
    const hasCallSignature = signatures.length > 0

    const signatureHasTypeParameters = (signature: ts.Signature) => {
      const typeParameterCount = signature.typeParameters?.length ?? 0

      return typeParameterCount > 0
    }

    const hasNoGenericSignature = !Array.some(signatures, signatureHasTypeParameters)
    const nonGenericCallableConditions = Array.make(hasCallSignature, hasNoGenericSignature)
    return Array.every(nonGenericCallableConditions, Boolean)
  }

  const initializerIsStructGet = (initializer: ts.Expression) => {
    const hasOneArgument = (call: ts.CallExpression) => strictEqual(1)(call.arguments.length)

    const symbolAtCalleeName = (callee: ts.PropertyAccessExpression) =>
      pipe(checker.getSymbolAtLocation(callee.name), Option.fromNullishOr)

    const resolveAlias = (symbol: ts.Symbol) => {
      const isAlias = (symbol.flags & ts.SymbolFlags.Alias) !== 0

      return isAlias ? checker.getAliasedSymbol(symbol) : symbol
    }

    const isGetName = flow(Struct.get<ts.Symbol, "name">("name"), strictEqual("get"))

    const structGetSymbol = (call: ts.CallExpression) =>
      pipe(
        call.expression,
        Option.liftPredicate(ts.isPropertyAccessExpression),
        Option.flatMap(symbolAtCalleeName),
        Option.map(resolveAlias),
        Option.filter(isGetName),
        Option.filter(symbolDeclaredInEffectStructModule)
      )

    return pipe(
      initializer,
      unwrapTransparentExpression,
      Option.liftPredicate(ts.isCallExpression),
      Option.filter(hasOneArgument),
      Option.flatMap(structGetSymbol),
      Option.isSome
    )
  }

  const matches = (declaration: ts.VariableDeclaration): ReadonlyArray<Detection> =>
    pipe(
      Option.gen(function* () {
        const localDeclaration = declarationIsExported(declaration)
          ? Option.none<ts.VariableDeclaration>()
          : Option.some(declaration)

        yield* localDeclaration
        const typeNode = yield* Option.fromNullishOr(declaration.type)
        const initializer = yield* Option.fromNullishOr(declaration.initializer)
        yield* Option.liftPredicate(typeNodeIsNonGenericCallable)(typeNode)
        yield* Option.liftPredicate(initializerIsStructGet)(initializer)

        return match({ node: typeNode, message, hint })
      }),
      Option.toArray
    )

  return matches
}

const variableDeclarationKinds = Array.of(ts.SyntaxKind.VariableDeclaration)

export const noMonomorphicStructGet = makeCheck(
  "no-monomorphic-struct-get",
  variableDeclarationKinds,
  ts.isVariableDeclaration,
  monomorphicStructGetMatches
)

File: internal/rules/no_monomorphic_struct_get/no_monomorphic_struct_get.go
package no_monomorphic_struct_get

import (
	"github.com/andrueandersoncs/better-typescript/internal/rule"
	"github.com/andrueandersoncs/typescript-go/ast"
	"github.com/andrueandersoncs/typescript-go/checker"
	"strings"
)

var message = rule.RuleMessage{Id: "no-monomorphic-struct-get", Description: "Avoid monomorphizing Struct.get at its declaration.", Help: "Keep Struct.get polymorphic. Inline it at a typed consumer, or put the domain type on the consuming value or result rather than on the getter."}
var Rule = rule.Rule{Name: "no-monomorphic-struct-get", Run: run}

func run(ctx rule.RuleContext, _ any) rule.RuleListeners {
	return rule.RuleListeners{ast.KindVariableDeclaration: func(node *ast.Node) {
		declaration := node.AsVariableDeclaration()
		if declaration.Type == nil || declaration.Initializer == nil {
			return
		}
		statement := node.Parent
		if statement != nil {
			statement = statement.Parent
		}
		if statement != nil && statement.Kind == ast.KindVariableStatement && ast.HasSyntacticModifier(statement, ast.ModifierFlagsExport) {
			return
		}
		declared := ctx.TypeChecker.GetTypeFromTypeNode(declaration.Type)
		signatures := ctx.TypeChecker.GetSignaturesOfType(declared, checker.SignatureKindCall)
		if len(signatures) == 0 {
			return
		}
		for _, signature := range signatures {
			if len(signature.TypeParameters()) > 0 {
				return
			}
		}
		initializer := unwrap(declaration.Initializer)
		if initializer.Kind != ast.KindCallExpression {
			return
		}
		call := initializer.AsCallExpression()
		if call.Arguments == nil || len(call.Arguments.Nodes) != 1 || call.Expression.Kind != ast.KindPropertyAccessExpression {
			return
		}
		access := call.Expression.AsPropertyAccessExpression()
		if access.Name().Text() != "get" {
			return
		}
		symbol := ctx.TypeChecker.GetSymbolAtLocation(access.Name())
		if symbol != nil && symbol.Flags&ast.SymbolFlagsAlias != 0 {
			symbol = ctx.TypeChecker.GetAliasedSymbol(symbol)
		}
		if symbol == nil {
			return
		}
		inEffect := false
		for _, item := range symbol.Declarations {
			if file := ast.GetSourceFileOfNode(item); file != nil {
				name := strings.ReplaceAll(file.FileName(), "\\", "/")
				if strings.Contains(name, "/effect/") && (strings.HasSuffix(name, "/Struct.d.ts") || strings.HasSuffix(name, "/Struct.ts")) {
					inEffect = true
				}
			}
		}
		if !inEffect {
			return
		}
		ctx.ReportNode(declaration.Type, message)
	}}
}
func unwrap(node *ast.Node) *ast.Node {
	for node != nil && (node.Kind == ast.KindParenthesizedExpression || node.Kind == ast.KindAsExpression || node.Kind == ast.KindSatisfiesExpression || node.Kind == ast.KindNonNullExpression) {
		node = node.Expression()
	}
	return node
}
